File: pset1/credit/credit.c
// check if the credit card number is valid

#include <stdio.h>
#include <cs50.h>

int main(void)
{
    long long number;
    long long numberCpy;
    int checksum = 0;

    // get the credit card number
    do
    {
        number = numberCpy = get_long_long("Number: ");
    }
    while (number <= 0);

    // calculate the checksum
    int i = 0;
    while (number > 1)
    {
        int sum = 0;

        if (i % 2 == 1)
        {
            // multiply every other digit by 2 starting with the penultimate digit
            sum = number % 10 * 2;
            if (sum >= 10)
            {
                sum = sum / 10 + sum % 10;
            }
        }
        else
        {
            // add the sum to the sum of the digits that weren't multiplied by 2
            sum = number % 10;
        }
        checksum += sum;
        number /= 10;

        i++;
    }

    // check the checksum
    if (checksum % 10 == 0)
    {
        // check the head numbers
        while (numberCpy > 10)
        {
            numberCpy /= 10;
        }

        switch (numberCpy)
        {
            case 3:
                printf("AMEX\n");
                break;
            case 5:
                printf("MASTERCARD\n");
                break;
            case 4:
                printf("VISA\n");
                break;
        }
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

File: pset6/credit/credit.py
from cs50 import get_string

# prompt user for credit card number
while True:
    n = get_string("Number: ")
    if (float(n) > 0):
        break

# calculate the checksum
checksum = 0
# the most significant number's index
index = len(n)

for c in n:
    # string to int
    c = int(c)
    # for even position
    if (index % 2 == 0):
        # if c * 2 > 10, add product's digits
        # else just add c * 2
        c = c * 2
        checksum += c // 10 + c % 10 if c >= 10 else c
    else:
        checksum += c
    index -= 1

# check checksum
# if last digit is 0, legit
if checksum % 10 == 0:
    # check head numbers
    if n.startswith(("34", "37")):
        print("AMEX")
    elif n.startswith("4"):
        print("VISA")
    elif n.startswith(("51", "52", "53", "54", "55")):
        print("MASTERCARD")
# not legit
else:
    print("INVALID")